valuable-derive/src/pointer.rs

```diff
@@ -0,0 +1,94 @@
+use std::collections::VecDeque;
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::parse::{Parse, ParseStream};
+use syn::{Expr, Result, Token};
+
+use crate::derive::respan;
+
+pub(crate) fn visit_pointer(input: TokenStream) -> Result<TokenStream> {
+    let Input {
+        expr,
+        segments,
+        visit,
+    } = syn::parse2(input)?;
+
+    let segments = segments.iter().map(|segment| match segment {
+        Segment::Member(syn::Member::Named(ident)) => {
+            let literal = ident.to_string();
+            quote! {
+                ::valuable::pointer::Segment::Field(#literal),
+            }
+        }
+        Segment::Member(syn::Member::Unnamed(index)) => {
+            quote! {
+                ::valuable::pointer::Segment::TupleIndex(#index),
+            }
+        }
+        Segment::Index(expr) => {
+            let expr = respan(quote! { &#expr }, expr);
+            quote! {
+                ::valuable::pointer::Segment::Index(
+                    ::valuable::Valuable::as_value(#expr)
+                ),
+            }
+        }
+    });
+
+    let visit_pointer = respan(quote! { ::valuable::Valuable::visit_pointer }, &expr);
+    Ok(quote! {
+        #visit_pointer(
+            &#expr,
+            ::valuable::pointer::Pointer::new(&[
+                #(#segments)*
+            ]),
+            &mut #visit,
+        )
```

Comment:
should we maybe also consider having a `pointer!` macro that returns a `Pointer` _without_ actually trying to visit it? the use-case i have in mind is storing a pointer in a struct so that it can be use to traverse multiple `Valuable`s.

Reply:
+1 on that, in my case I would like to create the pointer when parsing the templates where I don't have the data structures